site/docs/contributing.md

```diff
@@ -0,0 +1,316 @@
+
+We use ESLint and Prettier for code linting and formatting. Before submitting a pull request, please run:
+
+```sh
+npm run format
+npm run lint
+```
+
+It's a good idea to run the lint command as `npm run lint -- --fix` to automatically fix some linting errors.
+
+## Building the Project
+
+To build the project:
+
+```sh
+npm run build
+```
+
+For continuous building during development:
+
+```sh
+npm run build:watch
+```
+
+## Contributing to the CLI
+
+### Running the CLI During Development
+
+We recommend using `npm link` to link your local `promptfoo` package to the global `promptfoo` package:
+
+```sh
+npm link
+promptfoo --help
+```
+
+Alternatively, you can run the CLI directly:
+
+```sh
+npm run local -- eval --config examples/cloudflare-ai/chat_config.yaml
+```
+
+When working on a new feature, we recommend setting up a local `promptfooconfig.yaml` that tests your feature. Think of this as an end-to-end test for your feature.
+
+Here's a simple example:
+
+```yaml
+providers:
+  - name: openai:chat:gpt-4o
+prompts:
+  - Translate "{{input}}" to {{language}}
+tests:
+  - vars:
+      input: 'Hello, world!'
+      language: 'English'
+    assert:
+      - type: new-assertion-type
+```
+
+## Adding a New Provider
+
+1. Create an implementation in `src/providers/SOME_PROVIDER_FILE`.
+2. Update `loadApiProvider` in `src/providers.ts` to load your provider via string.
+3. Add test cases in `test/providers.test.ts`:
+   - Test the actual provider implementation.
+   - Test loading the provider via a `loadApiProvider` test.
+
+## Contributing to the Web UI
+
+The web UI is written as a [Next.js](https://nextjs.org/) app. It is exported as a static site and hosted by a local express server when bundled.
+
+To run the web UI in dev mode:
+
+```sh
+npm run local:web
+```
+
+This will host the web UI at http://localhost:3000. This allows you to hack on the Next.js app quickly (with fast refresh) but does not cover all features (such as running evals from a web browser) because that relies on a separate server process.
+
+To test the entire thing end-to-end, we recommend building the entire project and linking it to promptfoo:
+
+```sh
+npm run build
+npm link
+```
+
+Note that this will not update the web UI if you make further changes to the code. You have to run `npm run build` again.
+
+## Documentation
+
+If you're adding new features or changing existing ones, please update the relevant documentation. We use [Docusaurus](https://docusaurus.io/) for our documentation. We strongly encourage examples and guides as well.
+
+## Database
+
+promptfoo uses SQLite as its default database, managed through the Drizzle ORM. By default, the database is stored in `/.promptfoo/`. You can override this location by setting `PROMPTFOO_CONFIG_DIR`. The database schema is defined in `src/database.ts` and migrations are stored in `drizzle`. Note that the migrations are all generated and you should not access these files directly.
+
+### Main Tables
```

Comment:
these line-by-line property descriptions should be documented in code and I'm skeptical that it's worth reproducing here and having to update every time there's a migration - but fine for now